### PR TITLE
Add `trace-tools` crate

### PR DESCRIPTION
**Pull request by Adam-Gleave**
Adds the new `trace-tools` crate, providing functionality that should be useful for diagnostics.

## `trace-tools`

The crate consists of a few components:
 - The `Observe` trait
 - A `tracing::Subscriber` implementation, as well as useful `Layer`s.
 - Diagnostic utilities (currently just one, in `Flamegrapher`).
 - `#[observe]` attribute procedural macro.

## The `Observe` trait

This trait provides the ability to wrap futures with an instrumentation span, similar to the `tracing_futures::Instrument` trait. This span has some default properties that make it easy to register and filter in a subscriber:

```rust
tracing::trace_span!(
    target: "trace_tools::observe",
    "observed",
    observed.name = <my_function_name>,
    loc.file = <caller_file>,
    loc.line = <caller_line>,
    loc.col = <caller_column>,
);
```

Crucially, this span contains identical fields to those generated internally by `tokio` for task spawns (when compiled with `--cfg tokio_unstable`, which means that subscribers can interact with spans generated by `Observe` in the same way that they can with `tokio`s internal spans.

## Subscriber implementation

`trace-tools` provides a subscriber and two `Layer`s for dealing with spans and events:
 1) **Flamegraph** layer, produces a folded stack file detailing instrumented span stacks (either instrumented internally by `tokio` or by `trace-tools`) that can be used to generate a flamegraph of all *observed* code.
 2) **Logging** layer, allows `log` records to be converted into `tracing` events, and recreates full logging functionality (equivalent to `fern-logger`) within the subscriber itself, since it is impossible to set two loggers/subscribers at once. This means that logging remains consistent whether using `trace-tools` or not.

The subscriber can be initialised through a builder that can either set the global subscriber or return a `Layered` instance that can be further extended with more `Layer`s.

### Setting as the default subscriber:

```rust
// `Flamegrapher` handle returned, for building a flamegraph at the end of the run.
let flamegrapher = trace_tools::subscriber::build()
    .with_flamegraph_layer(stack_filename)
    .with_log_layer(log_config)
    .init()?
    .unwrap();
```

### Returning a `Layered` struct:

```rust
let (subscriber, flamegrapher) = trace_tools::subscriber::build()
    .with_flamegraph_layer(stack_filename)
    .with_log_layer(log_config)
    .finish()?;

// Extend the subscriber with external layers.
let subscriber = subscriber.with(console_layer);

// Set the global subscriber.
subscriber.init();
```

## `Flamegrapher`

Produces a flamegraph using the given folded stack file, using the `inferno` crate.

```rust
let flamegrapher = trace_tools::subscriber::build()
    .with_flamegraph_layer(stack_filename)
    .init()?
    .unwrap();

// Run some instrumented code...
// ...
// ...

// Programatically create the flamegraph file.
flamegrapher
    .with_graph_file("flamegraph.svg")?
    .write_flamegraph()?;
```

## `#[observe]` attribute

`trace-tools` provides a simple attribute proc-macro for instrumenting functions and futures with a span that specifies the `trace_tools::observe` target. The location fields in the span will describe the location of the tagged function or future, and the `observed.name` field will specify the function name:

```rust
use trace_tools::observe;

#[observe]
pub async fn say_hello() {
    println!("hello");
}
```

This is equivalent to the following:

```rust
trace_tools::Observe(say_hello(), "say_hello").await;
```

This macro can be used with regular, non-`async` functions too, unlike the `Observe` trait.

## Examples

There is an example for each layer in `trace-tools/examples`. The flamegraph example produces this interactive graph:

example

# More stuff to add

After the initial work is finished, it makes sense to also:
 - Add compatibility with tokio's [`console`](https://github.com/tokio-rs/console) project through another `Layer`: this will come in a subsequent PR.
 - Add some customisation options for the flamegraph (module paths, function names rather than locations, group stacks by thread, etc).

There are also plenty of things we can do with some additional subscriber `Layer`s if we ever wanted more diagnostic tools. In addition to being able to write our own, there are existing crates that provide layers for Chrome tracing and the Tracy profiler, and more.

### CODE REVIEW

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
You need to bump/release the logger to release the trace tools on crates.io

**Review by thibault-martinez — COMMENTED** [review]
Your PR description could be a nice readme for the crate...

File: fern-logger/src/config.rs [hunk 1]
```diff
@@ -63,6 +63,18 @@ impl LoggerOutputConfigBuilder {
         self
     }
 
+    /// Sets a collection of exclusions of a logger output.
+    /// A message is logged only if one of the exclusions is *not* part of the log's metadata target.
+    pub fn target_exclusions(mut self, target_exclusions: &[&str]) -> Self {
+        self.target_exclusions = Some(
+            target_exclusions
+                .iter()
+                .map(ToString::to_string)
```
Comment (thibault-martinez) [inline fern-logger/src/config.rs:72]: It triggers me that the impl is slightly different than `target_filters`, can you harmonise please ? To whatever you prefer.

Comment (Adam-Gleave) [reply fern-logger/src/config.rs:72]: Ha, yes that's fine :smile:

File: trace-tools/trace-tools/src/subscriber/layer/flamegraph.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+                    .span_locations
+                    .read()
+                    .get(&span.id())
+                    .and_then(|location| location.as_ref())
+                {
+                    Some(location) => location.to_string(),
+                    None => "unknown".to_string(),
+                };
+
+                write!(stack_str, "; {}", location).expect("writing to String should never fail");
```
Comment (grtlr) [inline trace-tools/trace-tools/src/subscriber/layer/flamegraph.rs:132]:
```suggestion
                write!(stack_str, "; {}", location).expect("writing to `String` should never fail");
```

**Review by grtlr — APPROVED** [review]
Looks great to me 👍. I only have a couple of bikeshedding comments.

File: trace-tools/trace-tools/examples/logging.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2021 IOTA Stiftung
```
[the remark this answers is not in the document]
Comment (Adam-Gleave) [reply trace-tools/trace-tools/examples/logging.rs:1]: Yeah definitely, I keep forgetting!

File: trace-tools/trace-tools/examples/logging.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    log();
+    filtered::log();
+    filtered::excluded::log();
+
+    Ok(())
+}
+
+pub fn log() {
+    log::info!("This should not log, it is not filtered");
```
[the remark this answers is not in the document]
Comment (Adam-Gleave) [reply trace-tools/trace-tools/examples/logging.rs:39]: The `filters` are used to opt in to logging events from the given targets, so I don't think that quite describes it -- it's the `exclusions` that are used to filter things out, and they're not being used for this target.

I've changed it to `This should not log, it is not captured by the filter`, which I think explains what's going on 🙂

File: fern-logger/CHANGELOG.md [hunk 5]
```diff
@@ -24,3 +24,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 
 - Initial features;
+
+## 0.2.0 - 2022-01-26
```
Comment (thibault-martinez) [inline fern-logger/CHANGELOG.md:28]: It goes above 😬

File: fern-logger/src/config.rs [hunk 6]
```diff
@@ -57,12 +57,24 @@ impl LoggerOutputConfigBuilder {
         self.target_filters = Some(
             target_filters
                 .iter()
-                .map(|f| f.to_string())
+                .map(ToString::to_string)
                 .collect::<Vec<String>>(),
```
Comment (thibault-martinez) [inline fern-logger/src/config.rs:61]:
```suggestion
                .collect::<Vec<_>>(),
```
For consistency

File: trace-tools/trace-tools/src/util/flamegrapher.rs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+impl Flamegrapher {
+    /// Creates a new [`Flamegrapher`] with no associated files.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns a [`Flamegrapher`] with the given folded stack file name.
+    ///
+    /// This file will be given the extension of `.folded`.
```
Comment (thibault-martinez) [inline trace-tools/trace-tools/src/util/flamegrapher.rs:28]: Does it actually do that ?

File: trace-tools/trace-tools/src/subscriber/mod.rs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+///  - This should only be called once. Any subsequent calls will fail, since the global logger can only be set
+/// once in a program's lifespan.
+///  - If the global logger has already been set (by the [`log`] crate, for example), this will fail.
+pub fn collect_logs() -> Result<(), log::SetLoggerError> {
+    LogTracer::init()
+}
+
+/// [`Layered`](tracing_subscriber::layer::Layered) type describing the composition of the subscriber constructed
+/// by the [`trace_tools`](crate) crate.
+pub type BeeSubscriber =
```
Comment (thibault-martinez) [inline trace-tools/trace-tools/src/subscriber/mod.rs:38]: Maybe rename this to something else now...

File: fern-logger/src/config.rs (hunk 6 above)
Comment (Adam-Gleave) [reply fern-logger/src/config.rs:61]: 🤦‍♂️

File: trace-tools/trace-tools/src/util/flamegrapher.rs (hunk 7 above)
Comment (Adam-Gleave) [reply trace-tools/trace-tools/src/util/flamegrapher.rs:28]: No, not anymore -- I'll update it.

**Review by thibault-martinez — APPROVED** [review]

File: fern-logger/src/config.rs [hunk 9]
```diff
@@ -57,8 +57,20 @@ impl LoggerOutputConfigBuilder {
...
-                .collect::<Vec<String>>(),
+                .map(ToString::to_string)
+                .collect::<Vec<_>>(),
+        );
+        self
+    }
+
+    /// Sets a collection of exclusions of a logger output.
+    /// A message is logged only if one of the exclusions is *not* part of the log's metadata target.
+    pub fn target_exclusions(mut self, target_exclusions: &[&str]) -> Self {
```
Comment (pvdrz) [inline fern-logger/src/config.rs:68]: shouldn't this method be `#[must_use]` due to the newer clippy lints?

File: trace-tools/trace-tools/src/subscriber/layer/flamegraph.rs [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+                {
+                    Some(location) => location.to_string(),
+                    None => "unknown".to_string(),
+                };
+
+                write!(stack_str, "; {}", location).expect("writing to `String` should never fail");
+            }
+        }
+
+        let _ = write!(stack_str, " {}", delta.as_micros());
```
Comment (pvdrz) [inline trace-tools/trace-tools/src/subscriber/layer/flamegraph.rs:136]: why not `expect` here too?

File: trace-tools/trace-tools/src/subscriber/layer/flamegraph.rs [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+impl Display for SpanLocation {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}:{}", self.file, self.line)
+    }
+}
+
+impl TryFrom<LocationVisitor> for SpanLocation {
+    // This is a bit odd, but since a `LocationVisitor` can fail to record any location information,
```
Comment (pvdrz) [inline trace-tools/trace-tools/src/subscriber/layer/flamegraph.rs:219]: I'd change this to `///` even though it is a private type.

File: trace-tools/trace-tools/src/subscriber/layer/log.rs [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+    fs::{File, OpenOptions},
+    io::{self, Stdout, StdoutLock},
+};
+
+/// Describes the output target of a [`log`] event.
+///
+/// Variants wrap a locked writer to the output target.
+enum LogOutput<'a> {
+    /// Log to standard output, with optional color.
+    Stdout(StdoutLock<'a>, bool),
```
Comment (pvdrz) [inline trace-tools/trace-tools/src/subscriber/layer/log.rs:29]: Insert joke here about `stdout` being accessed through a file descriptor

**Review by pvdrz — APPROVED** [review]

File: trace-tools/trace-tools/src/subscriber/layer/flamegraph.rs (hunk 10 above)
Comment (Adam-Gleave) [reply trace-tools/trace-tools/src/subscriber/layer/flamegraph.rs:136]: No idea -- will change it!

File: fern-logger/src/config.rs (hunk 9 above)
Comment (Adam-Gleave) [reply fern-logger/src/config.rs:68]: It probably should...weird that clippy isn't failing.

File: trace-tools/trace-tools/src/subscriber/layer/log.rs (hunk 12 above)
Comment (Adam-Gleave) [reply trace-tools/trace-tools/src/subscriber/layer/log.rs:29]: Good point :trollface:

File: fern-logger/src/config.rs (hunk 9 above)
Comment (Adam-Gleave) [reply fern-logger/src/config.rs:68]: Ah, it's because `#[must_use]` is applied to the entire `LoggerOutputConfigBuilder` at line 22, so this is fine.

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**